File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccess/Reauth.swift
import Foundation

/// The biometric primitive the re-auth gate drives. The real conformer wraps the
/// Secure-Enclave key-release (the released secret is zeroized + discarded — re-auth
/// only cares that the release succeeded). `isEnrolled` is the prompt-free predicate
/// that decides whether the gate engages at all.
///
/// `Sendable` because a `@MainActor` re-auth gate sends its conformer off-actor
/// to `await authorize` (#231).
public protocol BiometricAuthorizer: Sendable {
    var isEnrolled: Bool { get }
    /// Prove presence. Throws `DeviceUnlockError`-class failures on cancel / non-match
    /// / lockout. `async` because the real conformer drives an `LAContext` evaluation.
    func authorize(reason: String) async throws
}

/// A gate the view models `await` before each mutating write. Conformers decide
/// whether a write needs a fresh biometric prompt (grace window) and engage the
/// biometric only when required.
///
/// `Sendable` because a `@MainActor` view model sends its gate off-actor to
/// `await authorizeWrite` (#231).
public protocol WriteReauthGate: Sendable {
    /// Returns normally when the write may proceed (authorized, within the grace
    /// window, or not enrolled); throws when biometry was required and failed.
    func authorizeWrite(reason: String) async throws
}

/// v1 re-auth grace window. Writes inside this window after the last successful auth
/// do not re-prompt. One global value (no per-write-type tuning in v1).
///
/// This is only the fallback window for a `GraceWindowReauthGate` built without an
/// explicit `window` (e.g. host tests). The **iOS app's effective grace default is
/// 2 min**, not 30 s: the composition root seeds the shared `RetargetableReauthGate`
/// from the vault's persisted `reauth_grace_window_ms` setting, falling back to the
/// schema default `REAUTH_WINDOW_DEFAULT_MS` (2 min) when unset — see
/// `RetargetableGateFactory` in the app target.
public enum ReauthWindow {
    public static let v1Default: Duration = .seconds(30)
}

/// Pure policy: does a write need a fresh biometric prompt? `true` when never authed
/// (`lastAuthAt == nil`) or when at least `window` has elapsed since the last auth.
/// Boundary is inclusive: exactly `window` ⇒ re-auth required.
///
/// Times are `MonotonicInstant`s — a **monotonic** timeline, not wall-clock: the window
/// measures true elapsed time and must not move under an NTP correction or a user
/// clock-set, which on a wall clock can jump backward and silently extend the
/// silent-write window past `window` (issue #282). Only the gap between the two
/// instants is meaningful, so they must come from the same clock.
public func needsReauth(lastAuthAt: MonotonicInstant?, now: MonotonicInstant, window: Duration) -> Bool {
    guard let last = lastAuthAt else { return true }
    return last.duration(to: now) >= window
}

File: android/vault-access/src/main/kotlin/org/secretary/browse/Reauth.kt
package org.secretary.browse

/** Grace-window defaults for write re-authentication. Mirror of iOS `ReauthWindow.v1Default`. */
object ReauthWindow {
    /** Default seconds-as-millis a fresh presence proof stays valid before a write re-prompts.
     *  30 s matches the iOS biometric sibling; biometric prompts are fast, so a short window is
     *  low-friction. Not user-configurable in this slice (see the design doc, §2). */
    const val V1_DEFAULT_MS: Long = 30_000L
}

/**
 * Does a mutating write need a fresh presence proof?
 *
 * Pure policy — the single source of truth shared by every gate decision (host-tested in isolation).
 *
 * @param lastAuthAtMs epoch-millis of the last successful proof this session, or null if none yet.
 * @param nowMs        current epoch-millis.
 * @param windowMs     grace window; within it a write is silently authorized.
 * @return true if the user must re-prove presence:
 *   - `lastAuthAtMs == null`            → true  (never authed this session)
 *   - `nowMs - lastAuthAtMs >= window`  → true  (window elapsed; boundary INCLUSIVE)
 *   - otherwise                         → false (still inside the grace window)
 */
fun needsReauth(lastAuthAtMs: Long?, nowMs: Long, windowMs: Long): Boolean {
    if (lastAuthAtMs == null) return true
    return nowMs - lastAuthAtMs >= windowMs
}
